# Case 1: Climate Change in Regression Analysis

In [29]:
import pandas as pd
import statsmodels.formula.api as smf
import numpy as np

In [5]:
climate_train = pd.read_csv('https://raw.githubusercontent.com/amandeep0/IS451/main/data/ClimateChangeTrain.csv')
climate_test = pd.read_csv('https://raw.githubusercontent.com/amandeep0/IS451/main/data/ClimateChangeTest.csv')
climate_train.head()

,Year,Month,MEI,CO2,CH4,N2O,CFC11,CFC12,TSI,Aerosols,Temp
0,1983,5,2.556,345.96,1638.59,303.677,191.324,350.113,1366.1024,0.0863,0.109
1,1983,6,2.167,345.52,1633.71,303.746,192.057,351.848,1366.1208,0.0794,0.118
2,1983,7,1.741,344.15,1633.22,303.795,192.818,353.725,1366.2850,0.0731,0.137
3,1983,8,1.130,342.25,1631.35,303.839,193.602,355.633,1366.4202,0.0673,0.176
4,1983,9,0.428,340.17,1648.40,303.901,194.392,357.465,1366.2335,0.0619,0.149


## a) Linear Regression Model to predict Temp, using all variables, except Year and Month

In [7]:
linear_regression_model_1 = smf.ols('Temp ~ MEI + CO2 + CH4 + N2O + CFC11 + CFC12 + TSI + Aerosols', data=climate_train)
linear_regression_model_1_results = linear_regression_model_1.fit()
print(linear_regression_model_1_results.summary())

                            OLS Regression Results                            
Dep. Variable:                   Temp   R-squared:                       0.751
Model:                            OLS   Adj. R-squared:                  0.744
Method:                 Least Squares   F-statistic:                     103.6
Date:                Fri, 11 Oct 2024   Prob (F-statistic):           1.94e-78
Time:                        15:22:10   Log-Likelihood:                 280.10
No. Observations:                 284   AIC:                            -542.2
Df Residuals:                     275   BIC:                            -509.4
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -124.5943     19.887     -6.265      0.0

## a(i). Linear Regression Equation:

In [17]:
# Get intercept and coefficients
intercept = linear_regression_model_1_results.params['Intercept']
coefficients = linear_regression_model_1_results.params.drop('Intercept')

# Generate Linear Regression Equation
equation1 = f"Temp = {intercept:.4f} + " + " + ".join([f"{coef:.4f}({name})" for name, coef in coefficients.items()])
print("Linear Regression Equation:\n", equation1)

Linear Regression Equation:
 Temp = -124.5943 + 0.0642(MEI) + 0.0065(CO2) + 0.0001(CH4) + -0.0165(N2O) + -0.0066(CFC11) + 0.0038(CFC12) + 0.0931(TSI) + -1.5376(Aerosols)


## a(ii). Model Evaluation
$R^2$ = 0.751

Significant values (p-value < 0.05):
- MEI 
- CO2
- CFC11
- CHC12
- TSI
- Aerosols

## a(iii). Contradiction for N2O and CFC-11
The simplest explanation for this contradiction is due to multicollinearity that may be present with other variables (greehouse gasses). The model does not recognize individual effects towards other ones, and this results in the negative coefficient signs. 

## a(iv). Correlation between all Independent variables 


In [10]:
climate_train.corr()

,Year,Month,MEI,CO2,CH4,N2O,CFC11,CFC12,TSI,Aerosols,Temp
Year,1.000000,-0.027942,-0.036988,0.982749,0.915659,0.993845,0.569106,0.897012,0.170302,-0.345247,0.786797
Month,-0.027942,1.000000,0.000885,-0.106732,0.018569,0.013632,-0.013111,0.000675,-0.034606,0.014890,-0.099857
MEI,-0.036988,0.000885,1.000000,-0.041147,-0.033419,-0.050820,0.069000,0.008286,-0.154492,0.340238,0.172471
CO2,0.982749,-0.106732,-0.041147,1.000000,0.877280,0.976720,0.514060,0.852690,0.177429,-0.356155,0.788529
CH4,0.915659,0.018569,-0.033419,0.877280,1.000000,0.899839,0.779904,0.963616,0.245528,-0.267809,0.703255
N2O,0.993845,0.013632,-0.050820,0.976720,0.899839,1.000000,0.522477,0.867931,0.199757,-0.337055,0.778639
CFC11,0.569106,-0.013111,0.069000,0.514060,0.779904,0.522477,1.000000,0.868985,0.272046,-0.043921,0.407710
CFC12,0.897012,0.000675,0.008286,0.852690,0.963616,0.867931,0.868985,1.000000,0.255303,-0.225131,0.687558
TSI,0.170302,-0.034606,-0.154492,0.177429,0.245528,0.199757,0.272046,0.255303,1.000000,0.052117,0.243383
Aerosols,-0.345247,0.014890,0.340238,-0.356155,-0.267809,-0.337055,-0.043921,-0.225131,0.052117,1.000000,-0.384914


N2O is highly correlated to:
- CO2 
- CH4
- CFC12

CFC.11 is highly correlated to:
- CH4
- CFC12

Note: A variable is considered highly correlated if absolute correlation is higher than 0.7. 

## b) Linear Regression using only MEI, TSI, Aerosols, and N2O

In [37]:
linear_regression_model_2 = smf.ols('Temp ~ MEI + TSI + Aerosols + N2O', data=climate_train)
linear_regression_model_2_results = linear_regression_model_2.fit()
print(linear_regression_model_2_results.summary())

                            OLS Regression Results                            
Dep. Variable:                   Temp   R-squared:                       0.726
Model:                            OLS   Adj. R-squared:                  0.722
Method:                 Least Squares   F-statistic:                     184.9
Date:                Sun, 13 Oct 2024   Prob (F-statistic):           3.52e-77
Time:                        20:19:34   Log-Likelihood:                 266.64
No. Observations:                 284   AIC:                            -523.3
Df Residuals:                     279   BIC:                            -505.0
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -116.2269     20.223     -5.747      0.0

### b(i) Coefficient Comparison to Previous Model

In [38]:
### First Model
print("Previous Linear Regression Equation:\n", equation1)

### Second Model
# Get intercept and coefficients
intercept = linear_regression_model_2_results.params['Intercept']
coefficients = linear_regression_model_2_results.params.drop('Intercept')

# Generate Linear Regression Equation
equation2 = f"Temp = {intercept:.4f} + " + " + ".join([f"{coef:.4f}({name})" for name, coef in coefficients.items()])
print("Current Linear Regression Equation:\n", equation2)

Previous Linear Regression Equation:
 Temp = -124.5943 + 0.0642(MEI) + 0.0065(CO2) + 0.0001(CH4) + -0.0165(N2O) + -0.0066(CFC11) + 0.0038(CFC12) + 0.0931(TSI) + -1.5376(Aerosols)
Current Linear Regression Equation:
 Temp = -116.2269 + 0.0642(MEI) + 0.0795(TSI) + -1.7017(Aerosols) + 0.0253(N2O)


The coefficient for N2O for previous model is -0.0165 while the new model is 0.0253. The N20 coefficient in the previous model is negative, suggesting a negative relationship with temperature, while the new model is positive, suggesting positive relationship with temperature. If we look further in the coefficients from the p-value, the previous model suggest that N20 is not really significant variable while in the new model, N20 is highly significant due to 0 p-value. 

In conclusion, in the previous model, N20 has weak negative relationship with relationship with temperature while in the new model, N20 is strongly significant variable such that the higher N20 level, the higher temperature level. 

### b(ii) Quality Comparison to Previous Model

<b>Previous Model: </b> </br></br>

$R^2$ = 0.751</br>
Adjusted $R^2$ = 0.744

Significant values (p-value < 0.05):
- MEI 
- CO2
- CFC11
- CHC12
- TSI
- Aerosols

<b>New Model: </b> </br></br>
$R^2$ = 0.726</br>
Adjusted $R^2$ = 0.0.722

Significant values (p-value < 0.05):
- MEI 
- TSI
- N2O
- Aerosols

Comparing $R^2$ values, the previous model has a higher $R^2$ value than the new model, which means that the model contributes more towards the explanation in the variance of temperature. In terms of significance, the newer model's independent variables are all significant, which is not the case in the previous model. This means that the newer model is more reliable. In the previous model, it has been established that there may be multicollinearity present, and involved a lot of independent variables. The new model has fewer variables and less multicollinearity. 

## c) Test model

In [24]:
climate_test.head()

,Year,Month,MEI,CO2,CH4,N2O,CFC11,CFC12,TSI,Aerosols,Temp
0,2007,1,0.974,382.93,1799.66,320.561,248.372,539.206,1365.7173,0.0054,0.601
1,2007,2,0.510,383.81,1803.08,320.571,248.264,538.973,1365.7145,0.0051,0.498
2,2007,3,0.074,384.56,1803.10,320.548,247.997,538.811,1365.7544,0.0045,0.435
3,2007,4,-0.049,386.40,1802.11,320.518,247.574,538.586,1365.7228,0.0045,0.466
4,2007,5,0.183,386.58,1795.65,320.445,247.224,538.130,1365.6932,0.0041,0.372


In [39]:
prediction = linear_regression_model_2_results.predict(climate_test)
print(prediction)

0     0.504243
1     0.475001
2     0.450627
3     0.439461
4     0.450831
5     0.418853
6     0.422773
7     0.416541
8     0.370516
9     0.379791
10    0.376987
11    0.383793
12    0.398897
13    0.376185
14    0.355217
15    0.404700
16    0.444619
17    0.472767
18    0.462425
19    0.444292
20    0.422654
21    0.421001
22    0.438793
23    0.439419
dtype: float64


In [41]:
SSE = np.sum((prediction - climate_test['Temp'])**2)
SST = np.sum((climate_train['Temp'].mean() - climate_test['Temp'])**2)    

print("SSE:", SSE)
print("SST:", SST)

R2 = 1 - (SSE/SST)
print("R2:", R2)

SSE: 0.29489670825700687
SST: 0.5860188540964095
R2: 0.4967794872202327


$R^2$ on the test set is 0.4967794872202327, which is close to 0.5. Model is capturing moderate relationship between the independent variables and temperature. Still, more than 50% of variance in temperature could not be explained by this model, which means there are other factors that affects temperature. To improve this model, we can add more variables that are relevant, which may improve the R2 value. 